ethcore/src/trace/flat.rs

```diff
@@ -0,0 +1,183 @@
+// Copyright 2015, 2016 Ethcore (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+use trace::BlockTraces;
+use super::trace::{Trace, Action, Res};
+
+/// Trace localized in vector of traces produced by a single transaction.
+///
+/// Parent and children indexes refer to positions in this vector.
+pub struct FlatTrace {
+	/// Index of the parent trace within the same transaction.
+	pub parent: Option<usize>,
+	/// Indexes of child traces within the same transaction.
+	pub children: Vec<usize>,
+	/// VM depth.
+	pub depth: usize,
+	/// Type of action performed by a transaction.
+	pub action: Action,
+	/// Result of this action.
+	pub result: Res,
+}
+
+/// Represents all traces produced by a single transaction.
+pub struct FlatTransactionTraces(Vec<FlatTrace>);
+
+impl Into<Vec<FlatTrace>> for FlatTransactionTraces {
+	fn into(self) -> Vec<FlatTrace> {
+		self.0
+	}
+}
+
+/// Represents all traces produced by transactions in a single block.
+pub struct FlatBlockTraces(Vec<FlatTransactionTraces>);
+
+impl From<BlockTraces> for FlatBlockTraces {
+	fn from(block_traces: BlockTraces) -> Self {
+		let traces: Vec<Trace> = block_traces.into();
+		let ordered = traces.into_iter()
+			.map(|trace| FlatBlockTraces::flatten(None, 0, trace))
+			.map(FlatTransactionTraces)
+			.collect();
+		FlatBlockTraces(ordered)
+	}
+}
+
+impl Into<Vec<FlatTransactionTraces>> for FlatBlockTraces {
+	fn into(self) -> Vec<FlatTransactionTraces> {
+		self.0
+	}
+}
+
+impl FlatBlockTraces {
+	/// Helper function flattening nested tree structure to vector of ordered traces.
+	fn flatten(parent_index: Option<usize>, len: usize, trace: Trace) -> Vec<FlatTrace> {
+		let mut children = vec![];
+		let mut next_index = len + 1;
+		let all_subs = trace.subs
+			.into_iter()
+			.flat_map(|subtrace| {
+				let subs = FlatBlockTraces::flatten(Some(len), next_index, subtrace);
+				children.push(next_index);
+				next_index = next_index + subs.len();
+				subs
+			})
+			.collect::<Vec<_>>();
+
+		let ordered = FlatTrace {
```

Comment:
> probably want an "address" here, giving the exactly location in the call trace [index in root, index in first CALL, index in second CALL, ...] .

`FlatTrace` is not a part of public `trace` module api. In public api, there is another data structure called [`LocalizedTrace`](https://github.com/ethcore/parity/pull/997/files#diff-9e4ac8e6da1fc0d8eceebe2cb09f9fccR23). It has `trace_number`, `parent` (trace number of a parent) and `children` (trace numbers of children). Trace number is just a number of trace within given transaction. I think it should be sufficient.

> since depth is not enough to figure out sub-call order without an ordered depth-first enumeration of the entire trace, which may not be available (though i guess it could be done at time of retrieval rather than insertion).

Right now `flattening` is done at retrieval time.